File: java/Pipeline.java
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.ArrayBlockingQueue;
import javax.crypto.SecretKey;

public class Pipeline {
    public static Integer totalKeys;
    public static Queue<SecretKey> keyQueue;
    private static Integer numTransformationThreads;
    private static PipelineFactory factory;


    public static void loadConfiguration(PipelineFactory pipelineFactory) throws IOException{
        
        Pipeline.totalKeys = 10000;
        Pipeline.numTransformationThreads = 6;
        Pipeline.keyQueue = new ArrayBlockingQueue<>(totalKeys);
        Pipeline.factory = pipelineFactory;

        // I/O Files
        String inputPath = "./files/message.txt";
        TransformationTask.setInputFile(inputPath);
        String algorithm = Pipeline.factory.getAlgorithmName();
        String transformation = Pipeline.factory.getTransformationName();

        String outputPath = String.format("./java/output/%s_%s.txt", algorithm, transformation);
        TransformationTask.setOutputFile(outputPath);

        System.out.printf("Setting pipeline with %s\n", pipelineFactory.getClass().getName());
    }

    
    // Execution

    public static void run() throws Exception{
        System.out.println("Running Pipeline");
        long startTime = System.currentTimeMillis();
        
        // Inicio la generacion de claves
        KeyGenerationTask keyGenerationTask = Pipeline.factory.createKeyGenerationTask();
        Thread keyGenerationThread = new Thread(keyGenerationTask);
        keyGenerationThread.start();

        // Inicio las encripciones/desencripciones
        List<Thread> transformationThreads = new ArrayList<>();
        for(int i=0; i<Pipeline.numTransformationThreads; i++){
            TransformationTask transformationTask = Pipeline.factory.createTransformationTask();
            Thread transformationThread = new Thread(transformationTask);
            transformationThread.start();
            transformationThreads.add(transformationThread);
        }
        
        // Espero a que terminen
        keyGenerationThread.join();
        for(int i=0; i<Pipeline.numTransformationThreads; i++){
            transformationThreads.get(i).join();
        }
        TransformationTask.close_output_file();
        TransformationTask.resetKeyCounter();

        long endTime = System.currentTimeMillis();
        float executiontime = (endTime - startTime) / 1000F;
        System.out.printf("Pipeline took: --- %s seconds ---\n\n", executiontime);
    }
}

File: python/Pipeline.py

import threading
import queue
import time
from PipelineFactories import PipelineFactory
from KeyGenerationTasks import KeyGenerationTask
from TransformationTasks import TransformationTask

class Pipeline:

    def load_configuration(PipelineFactory:PipelineFactory):
        # Set up
        Pipeline.total_keys = 10000
        Pipeline.key_queue = queue.Queue()
        Pipeline.num_transformation_threads = 6
        TransformationTask.keys_counter = 0
        Pipeline.factory = PipelineFactory

        #  I/O Files
        input_path = './files/message.txt'
        TransformationTask.set_input_file(input_path)

        algorithm = PipelineFactory.algorithm
        transformation = PipelineFactory.transformation
        output_path = f'./python/output/{algorithm}_{transformation}.txt'
        TransformationTask.set_output_file(output_path)
        
        print(f'Setting up Pipeline with {type(PipelineFactory).__name__}')
        

    def run():
        print('Running Pipeline')
        start_time = time.time()

        key_generation_task = Pipeline.factory.create_key_generation_task()
        key_generation_thread = threading.Thread(target = key_generation_task.run)
        key_generation_thread.start()

        transformation_threads = []
        for i in range(Pipeline.num_transformation_threads):
            transformation_task = Pipeline.factory.create_transformation_task()
            transformation_thread = threading.Thread(target = transformation_task.run)
            transformation_thread.start()
            transformation_threads.append(transformation_thread)
        
        key_generation_thread.join()
        for thread in transformation_threads:
            thread.join()

        execution_time = round(time.time() - start_time, 4)
        print(f'Pipeline took: --- {execution_time} seconds ---\n')